File: src/llm-gateway/claude-cli.ts
/**
 * Thin runner around `claude -p --output-format stream-json`.
 *
 * Each completion spawns one CLI process with all agentic surface disabled
 * (`--tools ""`, no session persistence, no slash commands) and a caller-owned
 * system prompt, so the process behaves like a plain model call while reusing
 * whatever auth the CLI is already configured with (OAuth login or
 * ANTHROPIC_API_KEY).
 *
 * The prompt is written to stdin rather than passed as an argv entry so long
 * conversation transcripts never hit OS argument-length limits.
 */

import { spawn } from 'node:child_process';
import { createInterface } from 'node:readline';

export interface ClaudeUsage {
  inputTokens: number;
  outputTokens: number;
  cacheReadInputTokens: number;
  cacheCreationInputTokens: number;
}

export interface ClaudeCompletion {
  text: string;
  model: string;
  stopReason: string | null;
  usage: ClaudeUsage;
  costUsd: number | null;
}

export interface RunClaudeOptions {
  claudeBin: string;
  prompt: string;
  systemPrompt: string;
  model?: string;
  effort?: string;
  timeoutMs: number;
  signal?: AbortSignal;
  /** Called for each assistant text delta while the model is generating. */
  onTextDelta?: (text: string) => void;
}

export class ClaudeCliError extends Error {
  constructor(message: string, readonly statusCode: number = 502) {
    super(message);
    this.name = 'ClaudeCliError';
  }
}

interface StreamJsonLine {
  type: string;
  subtype?: string;
  message?: {
    model?: string;
    content?: Array<{ type: string; text?: string }>;
  };
  event?: {
    type: string;
    delta?: { type: string; text?: string; stop_reason?: string | null };
  };
  is_error?: boolean;
  result?: string;
  stop_reason?: string | null;
  total_cost_usd?: number;
  usage?: {
    input_tokens?: number;
    output_tokens?: number;
    cache_read_input_tokens?: number;
    cache_creation_input_tokens?: number;
  };
}

export function runClaude(options: RunClaudeOptions): Promise<ClaudeCompletion> {
  const args = [
    '-p',
    '--output-format', 'stream-json',
    '--include-partial-messages',
    '--verbose',
    '--tools', '',
    '--no-session-persistence',
    '--disable-slash-commands',
    '--system-prompt', options.systemPrompt,
  ];
  if (options.model) args.push('--model', options.model);
  if (options.effort) args.push('--effort', options.effort);

  return new Promise<ClaudeCompletion>((resolve, reject) => {
    const child = spawn(options.claudeBin, args, {
      stdio: ['pipe', 'pipe', 'pipe'],
      windowsHide: true,
    });

    let settled = false;
    let sawResult = false;
    let streamedText = '';
    let resultText: string | null = null;
    let model = options.model ?? 'unknown';
    let stopReason: string | null = null;
    let costUsd: number | null = null;
    const usage: ClaudeUsage = {
      inputTokens: 0,
      outputTokens: 0,
      cacheReadInputTokens: 0,
      cacheCreationInputTokens: 0,
    };
    let stderrTail = '';

    const finish = (outcome: { ok: true } | { ok: false; error: Error }) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      options.signal?.removeEventListener('abort', onAbort);
      if (child.exitCode === null) child.kill('SIGKILL');
      if (outcome.ok) {
        resolve({
          // The result line is authoritative; streamed deltas are the fallback
          // if the CLI ever omits it on success.
          text: resultText ?? streamedText,
          model,
          stopReason,
          usage,
          costUsd,
        });
      } else {
        reject(outcome.error);
      }
    };

    const onAbort = () => finish({ ok: false, error: new ClaudeCliError('request aborted by client', 499) });
    options.signal?.addEventListener('abort', onAbort, { once: true });
    if (options.signal?.aborted) onAbort();

    const timer = setTimeout(
      () => finish({ ok: false, error: new ClaudeCliError(`claude CLI timed out after ${options.timeoutMs}ms`, 504) }),
      options.timeoutMs,
    );

    child.on('error', (err) => {
      finish({ ok: false, error: new ClaudeCliError(`failed to spawn claude CLI: ${err.message}`, 500) });
    });

    child.stderr.on('data', (chunk: Buffer) => {
      stderrTail = (stderrTail + chunk.toString()).slice(-2000);
    });

    const rl = createInterface({ input: child.stdout });
    rl.on('line', (line) => {
      if (settled || !line.trim()) return;
      let parsed: StreamJsonLine;
      try {
        parsed = JSON.parse(line) as StreamJsonLine;
      } catch {
        return; // non-JSON noise on stdout — ignore
      }

      if (parsed.type === 'stream_event' && parsed.event) {
        const { event } = parsed;
        if (event.type === 'content_block_delta' && event.delta?.type === 'text_delta' && event.delta.text) {
          streamedText += event.delta.text;
          options.onTextDelta?.(event.delta.text);
        } else if (event.type === 'message_delta' && event.delta?.stop_reason) {
          stopReason = event.delta.stop_reason;
        }
        return;
      }

      if (parsed.type === 'assistant' && parsed.message?.model) {
        model = parsed.message.model;
        return;
      }

      if (parsed.type === 'result') {
        sawResult = true;
        if (parsed.is_error) {
          const detail = parsed.result || parsed.subtype || 'unknown error';
          finish({ ok: false, error: new ClaudeCliError(`claude CLI reported an error: ${detail}`) });
          return;
        }
        if (typeof parsed.result === 'string') resultText = parsed.result;
        if (parsed.stop_reason !== undefined) stopReason = parsed.stop_reason;
        if (typeof parsed.total_cost_usd === 'number') costUsd = parsed.total_cost_usd;
        if (parsed.usage) {
          usage.inputTokens = parsed.usage.input_tokens ?? 0;
          usage.outputTokens = parsed.usage.output_tokens ?? 0;
          usage.cacheReadInputTokens = parsed.usage.cache_read_input_tokens ?? 0;
          usage.cacheCreationInputTokens = parsed.usage.cache_creation_input_tokens ?? 0;
        }
      }
    });

    child.on('close', (code) => {
      if (settled) return;
      if (sawResult) {
        finish({ ok: true });
      } else {
        const detail = stderrTail.trim() || `exit code ${code}`;
        finish({ ok: false, error: new ClaudeCliError(`claude CLI exited without a result: ${detail}`) });
      }
    });

    child.stdin.write(options.prompt);
    child.stdin.end();
  });
}

File: internal/gateway/claudecli.go
package gateway

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"strings"
	"sync"
	"time"
)

type ClaudeCliError struct {
	Message    string
	StatusCode int
}

func (e *ClaudeCliError) Error() string {
	return e.Message
}

type streamJsonLine struct {
	Type         string   `json:"type"`
	Subtype      string   `json:"subtype"`
	IsError      bool     `json:"is_error"`
	Result       string   `json:"result"`
	StopReason   *string  `json:"stop_reason"`
	TotalCostUSD *float64 `json:"total_cost_usd"`
	Usage        *Usage   `json:"usage"`
	Message      *struct {
		Model   string `json:"model"`
		Content []struct {
			Type  string          `json:"type"`
			Text  string          `json:"text"`
			ID    string          `json:"id"`
			Name  string          `json:"name"`
			Input json.RawMessage `json:"input"`
		} `json:"content"`
	} `json:"message"`
	Event *struct {
		Type  string `json:"type"`
		Delta *struct {
			Type       string  `json:"type"`
			Text       string  `json:"text"`
			StopReason *string `json:"stop_reason"`
		} `json:"delta"`
		Usage *Usage `json:"usage"`
	} `json:"event"`
}

type RunClaudeOptions struct {
	ClaudeBin    string
	Prompt       string
	SystemPrompt string
	Model        string
	Effort       string
	TimeoutMs    int
	Signal       context.Context
	OnTextDelta  func(string)
}

func RunClaude(opts RunClaudeOptions) (*Completion, error) {
	ctx, cancel := context.WithTimeout(opts.Signal, time.Duration(opts.TimeoutMs)*time.Millisecond)
	defer cancel()

	args := []string{
		"-p",
		"--output-format", "stream-json",
		"--include-partial-messages",
		"--verbose",
		"--tools", "",
		"--no-session-persistence",
		"--disable-slash-commands",
		"--system-prompt", opts.SystemPrompt,
	}
	if opts.Model != "" {
		args = append(args, "--model", opts.Model)
	}
	if opts.Effort != "" {
		args = append(args, "--effort", opts.Effort)
	}

	cmd := exec.CommandContext(ctx, opts.ClaudeBin, args...)
	cmd.Stdin = strings.NewReader(opts.Prompt)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, &ClaudeCliError{
			Message:    fmt.Sprintf("failed to create stdout pipe: %v", err),
			StatusCode: 500,
		}
	}

	if _, err := cmd.StderrPipe(); err != nil {
		return nil, &ClaudeCliError{
			Message:    fmt.Sprintf("failed to create stderr pipe: %v", err),
			StatusCode: 500,
		}
	}

	if err := cmd.Start(); err != nil {
		return nil, &ClaudeCliError{
			Message:    fmt.Sprintf("failed to spawn claude CLI: %v", err),
			StatusCode: 500,
		}
	}

	var (
		streamedText string
		resultText   *string
		model        string
		stopReason   *string
		costUsd      *float64
		usage        Usage
		sawResult    bool
		stderrTail   strings.Builder
		settledOnce  sync.Once
		settled      bool
		settleMutex  sync.Mutex
	)

	model = opts.Model
	if model == "" {
		model = "unknown"
	}

	finish := func(completion *Completion, err error) (*Completion, error) {
		settleMutex.Lock()
		defer settleMutex.Unlock()

		settledOnce.Do(func() {
			settled = true
			if cmd.ProcessState == nil {
				_ = cmd.Process.Kill()
			}
		})

		if err != nil {
			return nil, err
		}
		return completion, nil
	}

	go func() {
		<-ctx.Done()
		if ctx.Err() == context.DeadlineExceeded {
			settleMutex.Lock()
			defer settleMutex.Unlock()
			if cmd.ProcessState == nil {
				_ = cmd.Process.Kill()
			}
		}
	}()

	scanner := bufio.NewScanner(stdout)
	for scanner.Scan() {
		settleMutex.Lock()
		if settled {
			settleMutex.Unlock()
			break
		}
		settleMutex.Unlock()

		line := scanner.Text()
		if line == "" {
			continue
		}

		var parsed streamJsonLine
		if err := json.Unmarshal([]byte(line), &parsed); err != nil {
			continue
		}

		if parsed.Type == "stream_event" && parsed.Event != nil {
			if parsed.Event.Type == "content_block_delta" && parsed.Event.Delta != nil {
				if parsed.Event.Delta.Type == "text_delta" && parsed.Event.Delta.Text != "" {
					streamedText += parsed.Event.Delta.Text
					if opts.OnTextDelta != nil {
						opts.OnTextDelta(parsed.Event.Delta.Text)
					}
				}
			} else if parsed.Event.Type == "message_delta" && parsed.Event.Delta != nil && parsed.Event.Delta.StopReason != nil {
				stopReason = parsed.Event.Delta.StopReason
			}
			continue
		}

		if parsed.Type == "assistant" && parsed.Message != nil && parsed.Message.Model != "" {
			model = parsed.Message.Model
			continue
		}

		if parsed.Type == "result" {
			sawResult = true
			if parsed.IsError {
				detail := parsed.Result
				if detail == "" {
					detail = parsed.Subtype
				}
				if detail == "" {
					detail = "unknown error"
				}
				_ = cmd.Wait()
				return finish(nil, &ClaudeCliError{
					Message:    fmt.Sprintf("claude CLI reported an error: %s", detail),
					StatusCode: 502,
				})
			}
			if parsed.Result != "" {
				resultText = &parsed.Result
			}
			if parsed.StopReason != nil {
				stopReason = parsed.StopReason
			}
			if parsed.TotalCostUSD != nil {
				costUsd = parsed.TotalCostUSD
			}
			if parsed.Usage != nil {
				usage = *parsed.Usage
			}
		}
	}

	_ = cmd.Wait()
	if ctx.Err() == context.DeadlineExceeded {
		return finish(nil, &ClaudeCliError{
			Message:    fmt.Sprintf("claude CLI timed out after %dms", opts.TimeoutMs),
			StatusCode: 504,
		})
	}

	if !sawResult {
		detail := fmt.Sprintf("exit code %d", cmd.ProcessState.ExitCode())
		if stderrTail.Len() > 0 {
			detail = stderrTail.String()
		}
		return finish(nil, &ClaudeCliError{
			Message:    fmt.Sprintf("claude CLI exited without a result: %s", detail),
			StatusCode: 502,
		})
	}

	text := streamedText
	if resultText != nil {
		text = *resultText
	}

	completion := &Completion{
		Text:       text,
		Model:      model,
		StopReason: stopReason,
		Usage:      usage,
		CostUSD:    costUsd,
	}

	return finish(completion, nil)
}

type SessionRunOptions struct {
	ClaudeBin     string
	Prompt        string
	SystemPrompt  string
	Model         string
	Effort        string
	MCPURL        string
	ToolTimeoutMS int
}

// mcpRelayServerName is the MCP server name the gateway registers itself
// under in the spawned CLI's --mcp-config. Claude Code exposes MCP tools to
// the model under the qualified name "mcp__<server>__<tool>" to avoid
// collisions with its own built-in tools, so any tool_use the model proposes
// arrives prefixed this way and must be stripped before it is handed back to
// an OpenAI-facing client, which only knows the bare name it declared.
const mcpRelayServerName = "relay"

func stripMCPToolPrefix(name string) string {
	prefix := "mcp__" + mcpRelayServerName + "__"
	return strings.TrimPrefix(name, prefix)
}

// StartClaudeSession spawns a tool-capable CLI process pointed at the
// gateway's own MCP endpoint for this session. The process is owned by the
// session table, not by any HTTP request: it survives the first response
// and is killed only when the session is reaped. Stream output is relayed
// onto sess.Events by a reader goroutine.
func StartClaudeSession(sm *SessionManager, sess *Session, opts SessionRunOptions) error {
	args := []string{
		"-p",
		"--output-format", "stream-json",
		"--include-partial-messages",
		"--verbose",
		"--tools", "",
		"--no-session-persistence",
		"--disable-slash-commands",
		"--system-prompt", opts.SystemPrompt,
		"--strict-mcp-config",
		"--mcp-config", fmt.Sprintf(`{"mcpServers":{%q:{"type":"http","url":"%s"}}}`, mcpRelayServerName, opts.MCPURL),
	}
	if opts.Model != "" {
		args = append(args, "--model", opts.Model)
	}
	if opts.Effort != "" {
		args = append(args, "--effort", opts.Effort)
	}

	cmd := exec.Command(opts.ClaudeBin, args...)
	cmd.Stdin = strings.NewReader(opts.Prompt)
	// The CLI's own MCP timeout must not fire before the gateway's tool
	// timeout does, or it would kill a legitimately parked call.
	mcpTimeout := opts.ToolTimeoutMS + 60000
	cmd.Env = append(os.Environ(),
		fmt.Sprintf("MCP_TOOL_TIMEOUT=%d", mcpTimeout),
		fmt.Sprintf("MCP_TIMEOUT=%d", mcpTimeout),
	)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return &ClaudeCliError{
			Message:    fmt.Sprintf("failed to create stdout pipe: %v", err),
			StatusCode: 500,
		}
	}

	if err := cmd.Start(); err != nil {
		return &ClaudeCliError{
			Message:    fmt.Sprintf("failed to spawn claude CLI: %v", err),
			StatusCode: 500,
		}
	}

	sess.Mu.Lock()
	sess.Cmd = cmd
	sess.Mu.Unlock()

	go relaySessionStream(sm, sess, stdout, cmd)
	return nil
}

func relaySessionStream(sm *SessionManager, sess *Session, stdout io.Reader, cmd *exec.Cmd) {
	defer close(sess.Events)

	scanner := bufio.NewScanner(stdout)
	scanner.Buffer(make([]byte, 0, 64*1024), 4*1024*1024)
	sawResult := false

	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}

		var parsed streamJsonLine
		if err := json.Unmarshal([]byte(line), &parsed); err != nil {
			continue
		}

		switch parsed.Type {
		case "stream_event":
			if parsed.Event == nil {
				continue
			}
			if parsed.Event.Type == "content_block_delta" && parsed.Event.Delta != nil &&
				parsed.Event.Delta.Type == "text_delta" && parsed.Event.Delta.Text != "" {
				sess.Events <- SessionEvent{Type: "text", Text: parsed.Event.Delta.Text}
			} else if parsed.Event.Type == "message_delta" {
				ev := SessionEvent{Type: "stop", Usage: parsed.Event.Usage}
				if parsed.Event.Delta != nil {
					ev.StopReason = parsed.Event.Delta.StopReason
				}
				sess.Events <- ev
			}
		case "assistant":
			if parsed.Message == nil {
				continue
			}
			var uses []ToolUseBlock
			for _, block := range parsed.Message.Content {
				if block.Type == "tool_use" {
					uses = append(uses, ToolUseBlock{ID: block.ID, Name: stripMCPToolPrefix(block.Name), Input: block.Input})
				}
			}
			ev := SessionEvent{Type: "tool_calls", Model: parsed.Message.Model}
			if len(uses) > 0 {
				// Register at announcement time so the CLI's MCP
				// tools/call can never race ahead of the pending table.
				ev.Calls = sm.RegisterAnnouncedCalls(sess, uses)
			}
			if ev.Model != "" || len(ev.Calls) > 0 {
				sess.Events <- ev
			}
		case "result":
			sawResult = true
			ev := SessionEvent{
				Type:       "run_end",
				StopReason: parsed.StopReason,
				CostUSD:    parsed.TotalCostUSD,
				Usage:      parsed.Usage,
			}
			if parsed.Result != "" {
				resultText := parsed.Result
				ev.ResultText = &resultText
			}
			if parsed.IsError {
				detail := parsed.Result
				if detail == "" {
					detail = parsed.Subtype
				}
				if detail == "" {
					detail = "unknown error"
				}
				ev.Err = &ClaudeCliError{
					Message:    fmt.Sprintf("claude CLI reported an error: %s", detail),
					StatusCode: 502,
				}
			}
			sess.Events <- ev
		}
	}

	_ = cmd.Wait()

	if !sawResult {
		sess.Events <- SessionEvent{
			Type: "run_end",
			Err: &ClaudeCliError{
				Message:    "claude CLI exited without a result",
				StatusCode: 502,
			},
		}
	}
}
